### PR TITLE
Remove type coercions from ScalarValue and aggregation function code

### PR DESCRIPTION
**Pull request by ozankabak**
# Which issue does this PR close?

Improves the situation on #2447.

 # Rationale for this change
There is an ongoing effort to reduce/eliminate type coercions in places like `ScalarValue` code, aggregation code and other downstream places (see #2355). The situation we want obtain is to move as many coercions as possible upstream to the extent that we can (see discussions in PR #3570).

This PR improves the current situation by sanitizing `ScalarValue` code and implementation of aggregation functions to remove the coercions therein. For the former, there is only one coercion left (in `try_from_value` function), the need for which arises due to an upstream hardcoding (which we will attempt to fix in the near future).

In many cases, this PR makes the code raise a compile-time error when performing "unsafe" operations that require a coercion (and forces the programmer to add the coercion explicitly). For example, one will get a compile time error if one tries to subtract an unsigned value from an uninitialized `ScalarValue`, or divide such a value to a decimal value -- this wasn't the case before.

# What changes are included in this PR?
As we were removing unnecessary coercions from the codebase, we encountered a "silent upcasting" behavior in the `sum` function, which promotes narrow types to 64-bit types *unconditionally*:

https://github.com/apache/arrow-datafusion/blob/57312284c082c914dd5e1edaa5c1fe3dbe4f222d/datafusion/physical-expr/src/aggregate/sum.rs#L217-L235

We are not 100% sure why this behavior was there, but removing it does not seem to result in any ill-effects. If anyone can tell us why such coercions are necessary, we can add them back in.

# Are there any user-facing changes?
No.

### CODE REVIEW

**Comment by ozankabak** [comment]
@alamb, this is one of our promised follow-up PR's to analyze `ScalarValue`-related coercions in the code and remove the ones we can.

**Comment by alamb** [comment]
> As we were removing unnecessary coercions from the codebase, we encountered a "silent upcasting" behavior in the sum function, which promotes narrow types to 64-bit types unconditionally:
>
> We are not 100% sure why this behavior was there, but removing it does not seem to result in any ill-effects. If anyone can tell us why such coercions are necessary, we can add them back in.

This is common when computing sums to avoid overflow, I think -- like the type of `SUM`ing `Int8` types can quickly end up as a `Int16` or larger. 

That being said, the fact there is no test is somewhat concerning

File: datafusion/physical-expr/src/aggregate/min_max.rs [hunk 1]
```diff
@@ -154,16 +154,10 @@ macro_rules! typed_min_max_batch_string {
...
 // Statically-typed version of min/max(array) -> ScalarValue for non-string types.
 macro_rules! typed_min_max_batch {
-    ($VALUES:expr, $ARRAYTYPE:ident, $SCALAR:ident, $OP:ident) => {{
-        let array = downcast_value!($VALUES, $ARRAYTYPE);
-        let value = compute::$OP(array);
-        ScalarValue::$SCALAR(value)
-    }};
-
-    ($VALUES:expr, $ARRAYTYPE:ident, $SCALAR:ident, $OP:ident, $TZ:expr) => {{
+    ($VALUES:expr, $ARRAYTYPE:ident, $SCALAR:ident, $OP:ident $(, $EXTRA_ARGS:ident)*) => {{
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/min_max.rs:157]: Macro 🧙

File: datafusion/physical-expr/src/aggregate/min_max.rs [hunk 2]
```diff
@@ -296,41 +290,18 @@ fn max_batch(values: &ArrayRef) -> Result<ScalarValue> {
...
-        ScalarValue::$SCALAR(match ($VALUE, $DELTA) {
-            (None, None) => None,
-            (Some(a), None) => Some(a.clone()),
-            (None, Some(b)) => Some(b.clone()),
-            (Some(a), Some(b)) => Some((*a).$OP(*b)),
-        })
-    }};
-
-    ($VALUE:expr, $DELTA:expr, $SCALAR:ident, $OP:ident, $TZ:expr) => {{
+    ($VALUE:expr, $DELTA:expr, $SCALAR:ident, $OP:ident $(, $EXTRA_ARGS:ident)*) => {{
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/min_max.rs:296]: these are drive by cleanups to reduce duplication in the macros, right?

(BTW if you submit these as individual free standing PRs you might find the reviews are faster -- finding enough contiguous time to review large PR changes can be challenging at times)

File: datafusion/physical-expr/src/aggregate/sum.rs [hunk 3]
```diff
@@ -206,87 +205,37 @@ pub(crate) fn sum_batch(values: &ArrayRef, sum_type: &DataType) -> Result<Scalar
...
-            sum_row!(index, accessor, rhs, f64)
-        }
-        (DataType::Float64, ScalarValue::UInt32(rhs)) => {
-            sum_row!(index, accessor, rhs, f64)
-        }
-        (DataType::Float64, ScalarValue::UInt16(rhs)) => {
-            sum_row!(index, accessor, rhs, f64)
-        }
-        (DataType::Float64, ScalarValue::UInt8(rhs)) => {
+    match s {
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/sum.rs:220]: I think this is a good change -- to use the type of the input to the type of the accumulator.

**Review by alamb — COMMENTED** [review]
Thank you very much for this PR. Very nice 🏅 I find it especially impressive that you have returned to help improve the codebase. 

The only thing I am slightly worried about is the change to casting in aggregates. I am going to write a quick test to make sure everything is fine

**Review by alamb — APPROVED** [review]
```rust
use std::sync::Arc;

use datafusion::arrow::array::{Int8Array};
use datafusion::arrow::record_batch::RecordBatch;
use datafusion::error::Result;
use datafusion::prelude::SessionContext;

/// This example demonstrates what happens summing a small int type
#[tokio::main]
async fn main() -> Result<()> {
    // define data that purposely will overflow an int8
    let array: Int8Array = (1..100).collect();
    let batch = RecordBatch::try_from_iter(vec![
        ("c", Arc::new(array) as _)
    ]).unwrap();

    let ctx = SessionContext::new();
    ctx.register_batch("t", batch).unwrap();

    println!("Count is:");
    ctx.sql("SELECT count(c) from t")
        .await
        .unwrap()
        .show()
        .await
        .unwrap();

    // Can't fit the sum in int8
    println!("Sum is:");
    ctx.sql("SELECT sum(c) from t")
        .await
        .unwrap()
        .show()
        .await
        .unwrap();

    Ok(())
}
```

It worked on both master and this branch 👍 

```
+------------+
| COUNT(t.c) |
+------------+
| 99         |
+------------+
Sum is:
+----------+
| SUM(t.c) |
+----------+
| 4950     |
+----------+
```

**Comment by alamb** [comment]
The runtime coercion in `ScalarValue` is pretty old, it might predate when we had better type coercion support in the optimizer. Maybe @andygrove  remembers

**Comment by ozankabak** [comment]
Yes, I did a bunch of drive-by cleanups while removing coercions :) The example you posted works because coercions are already done before we ever call `add_to_row`. My first thoughts regarding overflow were similar to yours, but then I realized by the time we perform additions we already have everything casted to wider types -- so the coercions I removed really seemed unnecessary.

**Comment by liukun4515** [comment]
I think the type coercion should be done in the `TypeCoercion` in the logical phase.
There are some TODO works for agg/buildin function/udf in the issue https://github.com/apache/arrow-datafusion/issues/3582#issuecomment-1254441358

**Comment by alamb** [comment]
> I think the type coercion should be done in the TypeCoercion in the logical phase.


I agree - and I think this PR is a step towards that goal (removes some unneeded logic in the physical phase as it is now done in the logical phase)

**Comment by alamb** [comment]
I will plan to merge this PR later today unless anyone else objects or would like more time to review

**Comment by alamb** [comment]
I merged this PR with master locally and ran the tests to ensure there are no logical conflicts. Looks good.

**Comment by alamb** [comment]
Thanks again @ozankabak

**Pull request merged**